Makes examples/benchmark optional and disables tests by default

nocko opened the pull request:
When using this project as a dependency in an application, having the tests, examples and benchmark built makes assumptions that aren't necessarily true about the target platform. Example: when building for embedded systems Posix type sockets aren't a feature and test/mocking frameworks aren't always available out-of-the-box.

It's convenient to disable these targets unless specifically developing on the library itself.

nocko committed:
Add options to disable examples and benchmarks

Makes examples/benchmark optional and disables tests by default

- Fix travis for optional test building
- Adds BUILD_EXAMPLES=ON to travis, since examples are tests too
- Fixes Typo in CMakeLists.txt

Reenable examples, benchmarks and tests by default

Removes unneeded .travis.yml changes

sijohans commented:
This project is also used in another project, in that project tests and examples are on by default. I.e., the logic is the opposite from this. Would it feel OK to have that approach instead?

Also: I remember that you asked about suggested BLE transport for this. This you find anything interesting?

nocko commented:
Hi there, sorry for the delay. 

I updated my patch to add the options, but leave them enabled by default. I reworked my project to explicitly disable them. I think the opposite logic may be more useful long-term, but for now this works for me.

During my absence I made a working salt-channel over BLE transport implementation in collaboration with a few others at AA. It seems to be working pretty well.

nocko committed:
Enable optimization in non Debug builds, optionally disable code coverage hooks

Preserve external CMAKE variables

nocko commented:
I added another patch that makes code coverage an option, it blows out the RAM on embedded systems. I also removed redundant \_DEBUG flags and have it preserve external CMAKE_* variables (needed for cross-compiling to work well).

I also think that this line:
 https://github.com/assaabloy-ppi/salt-channel-c/blob/780c02ad6c76d096cf11938cac6622078ce134d9/CMakeLists.txt#L35

prevented optimization of the code even in Release builds. I refactored this.

sijohans commented:
Hello, yes i think you are right about this. I never intended this build system to be complete in such way that it would work directly into any project. However, with changes like yours it is going towards that.

Is your project CMake based or do you just use this to cross compile the library? What cryptographic library are you using?

I'll merge this, thanks!

nocko commented:
My project is CMake based and uses libsodium. I initially tried tweetnacl, but it was terribly slow on the embedded system (~handshakes took 4.5 seconds).

I use salt-channel-c as a git submodule of my project. Here are the salt-channel specific lines of my CMakeLists.txt: 
``` cmake
add_definitions("-DUSE_SODIUM")
        
set(LIBSODIUM_LIBRARY sodium)
set(LIBSODIUM_ROOT_DIR ${PROJECT_SOURCE_DIR}/salt-channel-c/src/external/libsodium)

add_executable(${PROJECT_NAME}.elf ${NRF} ${SRC})

set(COVERAGE OFF CACHE BOOL "enable code coverage support")
set(BUILD_TESTS OFF CACHE BOOL "enable building tests in salt-channel")
set(BUILD_EXAMPLES OFF CACHE BOOL "enable building tests in salt-channel")
set(BUILD_BENCHMARK OFF CACHE BOOL "enable building tests in salt-channel")
add_subdirectory(salt-channel-c)

target_link_libraries(${PROJECT_NAME}.elf sodium sodium_wrapper)
set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -L${PROJECT_SOURCE_DIR}/salt-channel-c/src/external/libsodium/lib-arm")
```

Now that #44 is merged; I can use the upstream master branch directly (which is great, thanks!).

It's a little awkward to disable all the options before pulling in salt-channel-c, but hardly urgent.

sijohans commented:
I'll look into remaking the build system, I've learned a lot more about CMake since i started this project. I have some other thoughts about the read/write and time interface too, but i don't know when i can do this.

I assume you use nRF52, if you use the nRF52840 you could probably use the Arm CryptoCell 310. I seems to support most of the crypto primitives used for salt channel.
